dataset/src/jvmMain/kotlin/org/jetbrains/kotlinx/dl/dataset/preprocessor/image/Convert.kt

```diff
@@ -5,29 +5,31 @@
 
 package org.jetbrains.kotlinx.dl.dataset.preprocessor.image
 
+import org.jetbrains.kotlinx.dl.api.core.shape.TensorShape
 import org.jetbrains.kotlinx.dl.dataset.image.ColorMode
 import org.jetbrains.kotlinx.dl.dataset.image.colorMode
 import org.jetbrains.kotlinx.dl.dataset.image.imageType
-import org.jetbrains.kotlinx.dl.dataset.preprocessor.ImageShape
 import java.awt.image.BufferedImage
 
 /**
  * Converts provided image to the desired [ColorMode].
  *
  * @property [colorMode] target color mode.
  */
-public class Convert(public var colorMode: ColorMode = ColorMode.BGR) : ImagePreprocessorBase() {
-    override fun getOutputShape(inputShape: ImageShape): ImageShape {
-        return ImageShape(inputShape.width, inputShape.height, colorMode.channels.toLong())
-    }
-
-    override fun apply(image: BufferedImage): BufferedImage {
-        if (image.colorMode() == colorMode) return image
+public class Convert(public var colorMode: ColorMode = ColorMode.BGR) : ImageOperationBase() {
+    override fun apply(input: BufferedImage): BufferedImage {
+        if (input.colorMode() == colorMode) return input
         val outputType = colorMode.imageType()
-        val result = BufferedImage(image.width, image.height, outputType)
+        val result = BufferedImage(input.width, input.height, outputType)
         val graphics = result.createGraphics()
-        graphics.drawImage(image, 0, 0, null)
+        graphics.drawImage(input, 0, 0, null)
         graphics.dispose()
+
+        save?.save("convert_result", result)
```

Comment:
We have a bit of a problem when someone would try to save results of preprocessing multiple images, since they all will override each other. We should rethink how saving is done and if we should implement it differently.

Reply:
For example, we can add an operation which would allow to do anything with the result of the computation (so user would be able to save it whenever they want, or do something else useful):
```kotlin
public fun <I, O> Operation<I,O>.onResult(block: (O) -> Unit): Operation<I, O> {
    return PreprocessingPipeline<I, O, O>(this, object : Operation<O, O> {
        override fun apply(input: O): O {
            try {
                block(input)
            } finally {
                return input
            }
        }
        override fun getFinalShape(inputShape: TensorShape): TensorShape = inputShape
    })
}
```
Then to replicate the old behavior they would need to write something like:
```kotlin
private fun createDataLoader(preprocessedImagesDirectory: File): DataLoader<File> {
    return object : DataLoader<File> {
        override fun load(dataSource: File): Pair<FloatArray, TensorShape> {
            val preprocessing = pipeline<BufferedImage>()
                // do operations
                .resize {
                }.onResult { image ->
                    ImageIO.write(image, "jpg", File(preprocessedImagesDirectory, dataSource.name))
                }
                // do more operations
                .toFloatArray { }
            return preprocessing.apply(ImageConverter.toBufferedImage(dataSource))
        }
    }
}
```
On the other hand, we can just add a timestamp or some hash to the name of the image file and that would be enough.